### PR TITLE
Add autogenerated Tuple narrow methods for issue #1663

### PR DESCRIPTION
**Pull request by mfejzer**
Add autogenerated static Tuple narrow methods for issue #1663

### CODE REVIEW

**Comment by mfejzer** [comment]
Hello, I've added static narrow methods to Tuple, is it approach acceptable or should I add narrow directly to each TupleN classes?

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 1]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
 
+    /**
+     * Narrows a widened {@code Tuple1<? extends T1>} to {@code Tuple1<T1>}.
+     * This is eligible because immutable/read-only tuples are covariant.
+     * @param t A {@code Tuple1}.
+     * @param <T1> the 1st component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple1<T1>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1> Tuple1<T1> narrow(Tuple1<? extends T1> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 2]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
+    /**
+     * Narrows a widened {@code Tuple2<? extends T1, ? extends T2>} to {@code Tuple2<T1, T2>}.
+     * This is eligible because immutable/read-only tuples are covariant.
+     * @param t A {@code Tuple2}.
+     * @param <T1> the 1st component type
+     * @param <T2> the 2nd component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple2<T1, T2>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1, T2> Tuple2<T1, T2> narrow(Tuple2<? extends T1, ? extends T2> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 3]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
+     * Narrows a widened {@code Tuple3<? extends T1, ? extends T2, ? extends T3>} to {@code Tuple3<T1, T2, T3>}.
+     * This is eligible because immutable/read-only tuples are covariant.
+     * @param t A {@code Tuple3}.
+     * @param <T1> the 1st component type
+     * @param <T2> the 2nd component type
+     * @param <T3> the 3rd component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple3<T1, T2, T3>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1, T2, T3> Tuple3<T1, T2, T3> narrow(Tuple3<? extends T1, ? extends T2, ? extends T3> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T3' must match pattern '^\[A\-Z\]$'\.

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 4]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
+     * This is eligible because immutable/read-only tuples are covariant.
+     * @param t A {@code Tuple4}.
+     * @param <T1> the 1st component type
+     * @param <T2> the 2nd component type
+     * @param <T3> the 3rd component type
+     * @param <T4> the 4th component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple4<T1, T2, T3, T4>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1, T2, T3, T4> Tuple4<T1, T2, T3, T4> narrow(Tuple4<? extends T1, ? extends T2, ? extends T3, ? extends T4> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 5]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
+     * @param t A {@code Tuple5}.
+     * @param <T1> the 1st component type
+     * @param <T2> the 2nd component type
+     * @param <T3> the 3rd component type
+     * @param <T4> the 4th component type
+     * @param <T5> the 5th component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple5<T1, T2, T3, T4, T5>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1, T2, T3, T4, T5> Tuple5<T1, T2, T3, T4, T5> narrow(Tuple5<? extends T1, ? extends T2, ? extends T3, ? extends T4, ? extends T5> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 6]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
+     * @param <T1> the 1st component type
+     * @param <T2> the 2nd component type
+     * @param <T3> the 3rd component type
+     * @param <T4> the 4th component type
+     * @param <T5> the 5th component type
+     * @param <T6> the 6th component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple6<T1, T2, T3, T4, T5, T6>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1, T2, T3, T4, T5, T6> Tuple6<T1, T2, T3, T4, T5, T6> narrow(Tuple6<? extends T1, ? extends T2, ? extends T3, ? extends T4, ? extends T5, ? extends T6> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T3' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T6' must match pattern '^\[A\-Z\]$'\.

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 7]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
+     * @param <T2> the 2nd component type
+     * @param <T3> the 3rd component type
+     * @param <T4> the 4th component type
+     * @param <T5> the 5th component type
+     * @param <T6> the 6th component type
+     * @param <T7> the 7th component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple7<T1, T2, T3, T4, T5, T6, T7>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1, T2, T3, T4, T5, T6, T7> Tuple7<T1, T2, T3, T4, T5, T6, T7> narrow(Tuple7<? extends T1, ? extends T2, ? extends T3, ? extends T4, ? extends T5, ? extends T6, ? extends T7> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T4' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T6' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T7' must match pattern '^\[A\-Z\]$'\.

File: javaslang/src-gen/main/java/javaslang/Tuple.java [hunk 8]
```diff
@@ -344,4 +344,128 @@ static Tuple0 empty() {
...
+     * @param <T3> the 3rd component type
+     * @param <T4> the 4th component type
+     * @param <T5> the 5th component type
+     * @param <T6> the 6th component type
+     * @param <T7> the 7th component type
+     * @param <T8> the 8th component type
+     * @return the given {@code t} instance as narrowed type {@code Tuple8<T1, T2, T3, T4, T5, T6, T7, T8>}.
+     */
+    @SuppressWarnings("unchecked")
+    static <T1, T2, T3, T4, T5, T6, T7, T8> Tuple8<T1, T2, T3, T4, T5, T6, T7, T8> narrow(Tuple8<? extends T1, ? extends T2, ? extends T3, ? extends T4, ? extends T5, ? extends T6, ? extends T7, ? extends T8> t) {
```
Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T1' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T2' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T3' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T5' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T6' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T7' must match pattern '^\[A\-Z\]$'\.

Comment (sputnikci) [inline javaslang/src-gen/main/java/javaslang/Tuple.java]: \[Checkstyle\] ERROR: Name 'T8' must match pattern '^\[A\-Z\]$'\.

**Comment by danieldietrich** [comment]
Thank you, many (/most?) other static factory methods are also located in Tuple. I will take a look at the source later (I'm currently on a mobile device)

**Comment by danieldietrich** [comment]
Awesome, many thanks!

**Comment by mfejzer** [comment]
Thanks for review

**Pull request merged**